src/components/message/message.tsx

```diff
@@ -0,0 +1,199 @@
+import React, {
+  useCallback,
+  useContext,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'react'
+import { createPortal } from 'react-dom'
+import { CSSTransition, TransitionGroup } from 'react-transition-group'
+import {
+  CheckCircleOutline,
+  ExclamationCircleOutline,
+  InformationCircleOutline,
+  XCircleOutline,
+} from 'components/icons'
+import { ButtonSpinner } from 'components/button/button-spinner'
+import { MessageContext } from './message-context'
+
+export enum MessageTypes {
+  INFO = 'info',
+  SUCCESS = 'success',
+  WARNING = 'warning',
+  ERROR = 'error',
+  LOADING = 'loading',
+}
+
+export type MessageListType = {
+  id: string
+  title: string
+  type?: MessageTypes
+}
+
+/**
+ * MessageProvider properties
+ */
+type MessageProviderProps = {
+  /** wrap your App within MessageProvider component */
+  children: React.ReactElement
+  /** default time (in milli-seconds) after which Message component will disappear */
+  defaultDismissTime?: number
+}
+
+/**
+ * Display global messages as feedback in response to user operations.
+ *
+ * ### When to use
+ * * To provide feedback such as success, warning, error etc.
+ * * A message is displayed at top and center and will be dismissed automatically, as a non-interrupting light-weighted prompt.
+ * * Mostly useful to notify user for success or error API response
+ *
+ * ### How to use
+ *
+ * * Wrap your `<App />` within `<MessageProvider />` component
+ * * Import `useMessage` hook wherever you want to render `Message` component
+ * * `useMessage` provides you 2 methods which you can access using `const { message, removeMessage } = useMessage()`
+ *
+ * There are 5 different types of messages available to user. You can pass the string which you want to render in `Message` component as argument.
+ *
+ * * `message.info()`
+ * * `message.success()`
+ * * `message.error()`
+ * * `message.warning()`
+ * * `message.loading()`
+ *
+ * Use `removeMessage()` method to programmatically remove a message. `removeMessage()` takes the id of `Message` component you want to remove.
+ * All `message` methods return the id of the message which you can use with `removeMessage()`
+ */
+
+export function MessageProvider({
+  children,
+  defaultDismissTime = 3000,
+}: MessageProviderProps) {
+  const toastContainer = useRef<HTMLDivElement>(document.createElement('div'))
+
+  useLayoutEffect(() => {
+    const container = toastContainer.current
+    document.body.appendChild(container)
+    return () => {
+      document.body.removeChild(container)
+    }
+  }, [])
+
+  const [messages, setMessages] = useState<MessageListType[]>([])
+
+  const removeMessage = (id: string) => {
+    setMessages((prevState) => {
+      if (prevState.find((message) => message.id === id)) {
+        return prevState.filter((message) => message.id !== id)
+      }
+      return prevState
+    })
+  }
+
+  const dispatchMessage = (type: MessageTypes) => (
+    title: string,
+    dismissTime?: number,
```

Comment:
Just like `title` we should also provide a way for the user to send custom icon. Instead of only passing `dismissTime`, we can allow an `options` object.

```ts
type options = {
  dismissTime: number,
  icon?: React.Element
}
```

If the user doesn't provide any icon we would fallback to the default icon.